libraries/GCS_MAVLink/GCS.h

```diff
@@ -107,6 +109,189 @@ enum ap_message : uint8_t {
+    // Mission_Command is used as an intermediate form for both of
+    // them:
+    void handle_mission_item(const mavlink_message_t &msg,
+                             AP_Mission::Mission_Command &cmd);
+
+    void handle_mission_clear_all(const GCS_MAVLINK &link,
+                                  const mavlink_message_t &msg);
+
+    void queued_request_send();
+    void update();
+
+    bool active_link_is(const GCS_MAVLINK *_link) const { return _link == link; };
+
+    virtual MAV_MISSION_TYPE mission_type() const = 0;
+
+    bool receiving; // currently sending requests and expecting items
+
+protected:
+
+    GCS_MAVLINK *link; // link currently receiving waypoints on
+
+    // return the ap_message which can be queued to be sent to send a
+    // item request to the GCS:
+    virtual ap_message next_item_ap_message_id() const = 0;
+
+    void handle_mission_clear_all();
+
+    virtual bool clear_all_items() = 0;
+
+private:
+
+    virtual void _handle_mission_count(class GCS_MAVLINK &link,
+                                       const mavlink_mission_count_t &packet,
+                                       const mavlink_message_t &msg) = 0;
+
+    uint16_t        request_i; // request index
+    uint16_t        request_last; // last request index
+
+    // waypoints
+    uint8_t        dest_sysid; // where to send requests
+    uint8_t        dest_compid; // "
+    uint32_t        timelast_receive_ms;
+    uint32_t        timelast_request_ms;
+    const uint16_t  upload_timeout_ms = 8000;
+
+    // support for GCS getting waypoints etc from us:
+    virtual MAV_MISSION_RESULT _handle_mission_request_int(const mavlink_mission_request_int_t &packet,
+                                                           mavlink_mission_item_int_t &ret_packet) = 0;
+
+    virtual MAV_MISSION_RESULT _handle_mission_request(
+        const GCS_MAVLINK &_link,
+        const mavlink_message_t &msg,
+        const mavlink_mission_request_t &packet,
+        mavlink_mission_item_t &ret_packet) = 0;
+
+    void init_send_requests(GCS_MAVLINK &_link,
+                            const mavlink_message_t &msg,
+                            const int16_t _request_first,
+                            const int16_t _request_last);
+
+    void send_mission_ack(const mavlink_message_t &msg, MAV_MISSION_RESULT result) const;
+    void send_mission_ack(const GCS_MAVLINK &link, const mavlink_message_t &msg, MAV_MISSION_RESULT result) const;
+
+    virtual uint16_t item_count() const = 0;
+    virtual uint16_t max_items() const = 0;
+
+    virtual MAV_MISSION_RESULT replace_cmd(const AP_Mission::Mission_Command &cmd) = 0;
+    virtual MAV_MISSION_RESULT add_cmd(AP_Mission::Mission_Command &cmd) = 0;
+
+    virtual void complete(const GCS_MAVLINK &_link) {};
+
+};
+
+class MissionItemProtocol_Waypoints : public MissionItemProtocol {
+public:
+    MissionItemProtocol_Waypoints(AP_Mission &_mission) :
+        mission(_mission) {}
```

Comment:
You don't actually need to pass _mission in, you can use the singleton at runtime to fetch it.

Reply:
Yes indeed, and I did consider doing that, and I could change to do that.  The chunk of code which instantiates one of these then looks a bit odd (because I did try it that way) because it is checking for the non-nullness of something and then not doing anything with that non-null value.

Reply:
The init code is actually why I flagged this. It's just messy as is?